pkg/compact/schedule_delete.go

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package compact
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/block"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"github.com/thanos-io/thanos/pkg/runutil"
+)
+
+// ScheduleBlockDelete marks the block to be deleted.
+type ScheduleBlockDelete struct {
+	dir         string
+	logger      log.Logger
+	deleteDelay time.Duration
+	bkt         objstore.Bucket
+}
+
+// NewScheduleBlockDelete creates a new ScheduleBlockDelete.
+func NewScheduleBlockDelete(logger log.Logger, dir string, bkt objstore.Bucket, deleteDelay time.Duration) *ScheduleBlockDelete {
+	return &ScheduleBlockDelete{
+		dir:         dir,
+		logger:      logger,
+		deleteDelay: deleteDelay,
+		bkt:         bkt,
+	}
+}
+
+// ScheduleDelete deletes blocks from bucket
+// deleteDelay duration after block is marked for deletion.
+func (s *ScheduleBlockDelete) ScheduleDelete(ctx context.Context) error {
+	defer func() {
+		if err := os.RemoveAll(s.dir); err != nil {
+			level.Error(s.logger).Log("msg", "failed to remove compactor meta dir", s.dir, "err", err)
+		}
+	}()
+
+	if err := os.RemoveAll(s.dir); err != nil {
+		return errors.Wrap(err, "clean compaction compactor meta dir")
+	}
+
+	return s.bkt.Iter(ctx, "", func(name string) error {
+		id, ok := block.IsBlockDir(name)
+		if ok {
+			compactorMetaFile := path.Join(id.String(), metadata.CompactorMetaFilename)
+
+			r, err := s.bkt.Get(ctx, compactorMetaFile)
+			if s.bkt.IsObjNotFoundErr(err) {
+				return nil
+			}
+			if err != nil {
+				return errors.Wrapf(err, "get compactor-meta file: %v", compactorMetaFile)
```

Comment:
I'd suggest to use 
```suggestion
				return errors.Wrapf(err, "get file: %v", compactorMetaFile)
```

since `compactorMetaFile` already has `compactor-meta.json` in it. (Also at other places)